### PR TITLE
Fixes hydroponics trays wasting water (take two)

### PR DESCRIPTION
**Pull request by SuperSayu**
## About The Pull Request

Cleaner implementation of my last attempt at this bugfix.  Moves hydroponics plumbing code into its own file instead of being stuck in the middle of the machine logic.  Still handles water and nutrients separately through the duct network, so you won't waste one when looking for the other.

When using the duct system to feed in nutrients, it will only feed up to 15/20 of the nutrient tray, so that you can use mutagen, pest spray, etc, on trays as necessary.  This is a var that can be set in code, but I didn't set up a way to change it in game.

image
A setup like this will Just Work(tm).  All hydro trays will be maintained at 100 water until the sources run out.  You can dump nutrient into the input gate and it will spread through the system.  With just these two ingredients plus an infinite supply of seeds you too can have Too Much Food(tm).

## Why It's Good For The Game

Fix.

## Changelog

:cl:
fix: Using ducts to connect a water tank to one or more hydroponics trays is no longer a monumental waste of resources.
fix: Hydroponics trays fed nutrients through a duct will leave room for you to add chemicals by hand.
/:cl:

### CODE REVIEW

File: code/datums/components/plumbing/_plumbing.dm [hunk 1]
```diff
@@ -124,14 +124,18 @@
 	else if(reagents.total_volume > 0) //take whatever
 		return TRUE
 
+///returns TRUE if the source's reagents has only one entry
+/datum/component/plumbing/proc/is_pure_source()
+	return reagents.reagent_list.len == 1
```
Comment (ZephyrTFA) [inline code/datums/components/plumbing/_plumbing.dm:129]:
```suggestion
	return length(reagents.reagent_list) == 1
```
Supposedly length(X) is better than X.len

File: code/modules/hydroponics/hydroponics_plumbing.dm [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	return ..()
+
+// Add a check for the water level in determining whether or not to request fluids.
+/datum/component/plumbing/hydroponics/process()
+	if(!demand_connects || !reagents)
+		STOP_PROCESSING(SSplumbing, src)
+		return
+	var/obj/machinery/hydroponics/constructable/hydro_parent = parent
+	if(hydro_parent.reagents.total_volume < hydro_parent.maxductnutri || hydro_parent.waterlevel < hydro_parent.maxwater)
+		for(var/D in GLOB.cardinals)
```
Comment (ZephyrTFA) [inline code/modules/hydroponics/hydroponics_plumbing.dm:36]: use dir over D

File: code/modules/hydroponics/hydroponics_plumbing.dm [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+			if(D & demand_connects)
+				send_request(D)
+
+/// split_request_across(list/sources,amt,datum/ductnet/net): Request chems selectively from the given sources
+/// Returns: The amount transferred
+/datum/component/plumbing/hydroponics/proc/split_request_across(list/sources, amt, datum/ductnet/net)
+	var/suppliersLeft = sources.len
```
Comment (ZephyrTFA) [inline code/modules/hydroponics/hydroponics_plumbing.dm:43]: snake_case please

**Comment by Mothblocks** [comment]
As we have an abnormally large number of open PRs, I am closing all ones that have been marked stale. Please find an interested maintainer over Discord to reopen and merge the PR.

**Pull request closed without merging**